packages/button/src/Button.tsx

```diff
@@ -1,483 +1,182 @@
-  transition: all 120ms ease;
-  user-select: none;
-  padding: 0;
-
-  &:focus {
-    outline: none;
-  }
-
-  &:hover {
-    text-decoration: none;
-  }
-
-  // We're using CSS pseudo elements here in order to
-  // transition the gradients between button pseudo classes.
-  &:before,
-  &:after {
-    content: '';
-    transition: 0.15s opacity ease-in-out;
-    opacity: 0;
-    position: absolute;
-    top: 0;
-    bottom: 0;
-    left: 0;
-    right: 0;
-    border-radius: 2px;
-  }
-
-  &:not(:disabled) {
-    &:focus,
-    &:hover {
-      &:before {
-        opacity: 1;
-      }
-    }
+const iconSpacing = {
+  [Size.XSmall]: 6,
+  [Size.Small]: 6,
+  [Size.Default]: 6,
+  [Size.Large]: spacing[2],
+};
 
-    &:active:after {
-      opacity: 1;
-    }
+const Button: ExtendableBox<
+  ButtonProps & { ref?: React.Ref<any> },
+  'button'
+> = React.forwardRef(function Button(
+  {
+    variant = Variant.Default,
+    size = Size.Default,
+    darkMode = false,
+    baseFontSize = 14,
+    disabled = false,
+    leftGlyph,
+    rightGlyph,
+    children,
+    className,
+    ...rest
+  }: ButtonProps,
+  forwardRef,
+) {
+  const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+  const rippleRef = useRef<HTMLDivElement | null>(null);
+
+  useEffect(() => {
+    if (rippleRef.current != null) {
+      registerRipple(rippleRef.current, { variant, darkMode });
+    }
+  }, [rippleRef, variant, darkMode]);
+
+  const isIconOnlyButton = ((leftGlyph || rightGlyph) && !children) ?? false;
+
+  const buttonClassName = getClassName({
+    variant,
+    size,
+    darkMode,
+    baseFontSize,
+    disabled,
+    showFocus,
+  });
+
+  const isAnchor = typeof rest.href === 'string';
+  let type: JSX.IntrinsicElements['button']['type'];
+
+  // @ts-expect-error rest.as may be defined
+  if ((rest.as && rest.as === 'button') || (!isAnchor && !rest.as)) {
+    type = 'button';
   }
-`;
-
-const disabledStyle = css`
-  color: ${uiColors.gray.base};
-  border-color: ${uiColors.gray.light1};
-  background-color: ${uiColors.gray.light2};
-  background-image: none;
-  box-shadow: none;
-  pointer-events: none;
-`;
 
-interface BaseButtonProps {
-  disabled?: boolean;
-  darkMode?: boolean;
-  variant?: Variant;
-  size?: Size;
-  glyph?: React.ReactElement;
-  className?: string;
-  borderRadius?: string;
-  children?: React.ReactNode;
-  href?: string;
-  forceState?: {
-    focused?: boolean;
-    active?: boolean;
-    // Supporting "hovered" requires factoring out all the &:hover
-    // styles from the the base style and each variant's styles
+  const buttonProps = {
+    type,
+    className: cx(buttonClassName, className),
+    ref: forwardRef,
+    // Provide a default value for the as prop
+    // If consumping application passes a value for as, it will override the default set here
+    as: isAnchor ? 'a' : 'button',
+    // only add a disabled prop if not an anchor
+    ...(typeof rest.href !== 'string' && { disabled }),
```

Comment:
This looks like it would error if `typeof rest.href !== 'string` evaluates to false.

Reply:
what do you mean by error? This just only sets the disabled attribute when the component isn't an anchor because a tags aren't supposed to have disabled set, just aria-disabled.

Reply:
afaict this isn't causing any errors (and this is recycled code from our old button implementation)

Reply:
Is `...(false)` valid? 🤔

Reply:
I don't see that in the HTML

Reply:
I wasn't aware that JavaScript will just ignore it when you spread a boolean. I would've thought it would give you a type error. Learn something new every day.

Reply:
(Carry on)